components/Headlines.tsx

```diff
@@ -1,66 +1,110 @@
+type HeadlineProps = {
+  children: string | React.ReactNode[];
+  attributes?: Record<string, any>;
+};
 
-export const Headline1 = ({ children, attributes }: HeadlineProps) => <Headline Tag={Headline1Tag} attributes={attributes}>{children}</Headline>
-export const Headline2 = ({ children, attributes }: HeadlineProps) => <Headline Tag={Headline2Tag} attributes={attributes}>{children}</Headline>
-export const Headline3 = ({ children, attributes }: HeadlineProps) => <Headline Tag={Headline3Tag} attributes={attributes}>{children}</Headline>
-export const Headline4 = ({ children, attributes }: HeadlineProps) => <Headline Tag={Headline4Tag} attributes={attributes}>{children}</Headline>
+export const Headline1 = ({ children, attributes }: HeadlineProps) => (
+  <Headline Tag={Headline1Tag} attributes={attributes}>
+    {children}
+  </Headline>
+);
+export const Headline2 = ({ children, attributes }: HeadlineProps) => (
+  <Headline Tag={Headline2Tag} attributes={attributes}>
+    {children}
+  </Headline>
+);
+export const Headline3 = ({ children, attributes }: HeadlineProps) => (
+  <Headline Tag={Headline3Tag} attributes={attributes}>
+    {children}
+  </Headline>
+);
+export const Headline4 = ({ children, attributes }: HeadlineProps) => (
+  <Headline Tag={Headline4Tag} attributes={attributes}>
+    {children}
+  </Headline>
+);
 
-const Headline = ({ children, Tag, attributes: propAttributes }: {
-  children: string | React.ReactNode[]
-  Tag: React.FunctionComponent<TagProps>
-  attributes?: Record<string, any>
+const Headline = ({
+  children,
+  Tag,
+  attributes: propAttributes,
+}: {
+  children: string | React.ReactNode[];
+  Tag: React.FunctionComponent<TagProps>;
+  attributes?: Record<string, any>;
 }) => {
-  const router = useRouter()
-  const asPath = router.asPath
-  const slug = slugifyMarkdownHeadline(children as any[])
+  const router = useRouter();
+  const [isActive, setIsActive] = useState(false);
+  const slug = slugifyMarkdownHeadline(children as any);
+  const asPath = router.asPath;
+
+  useEffect(() => {
+    const hashIndex = asPath.indexOf('#');
+    const slugFromPath = hashIndex !== -1 ? asPath.slice(hashIndex + 1) : null;
+
+    setIsActive(slug === slugFromPath);
+
+  }, [router.asPath]);
+
+  const handleHeadingClick = () => {
+    const url = new URL(asPath, HOST);
+    const newHash = `#${slug}`;
+    url.hash = newHash;
+
+    const urlString = url.toString().substr(HOST.length, Infinity);
+    router.push(urlString, undefined, { shallow: true });
+    setIsActive(true); // Set isActive to true after clicking
+  };
 
   const attributes = {
     ...propAttributes,
     id: propAttributes?.slug || slug,
-    className: classnames('group cursor-pointer hover:underline', propAttributes?.className),
-    onClick: () => {
-      const url = new URL(asPath, HOST)
-      // recalculation necessary because of scope issue
-      const slug = slugifyMarkdownHeadline(children as any[])
```

Comment:
Here, the previous implementation states 
> // recalculation necessary because of scope issue

This PR changes it, why is it no longer necessary